4-star-nosed_mole/multipart_upload_specifications.md

```diff
@@ -0,0 +1,108 @@
+# Up-/Download Client & Production POC (Star-Nosed Mole) - Multipart Uploads
+
+This is an addition to the Star-Nosed mole epic to adress challenges in multipart uploads
+
+## Scope:
+A scope definition can be found here: https://wiki.verbis.dkfz.de/x/fAFzCQ
+
+## User Journeys
+
+The addition to this epic covers the following user journeys:
+
+### Multipart Data Upload:
+![Data Upload](./images/multipart_data_upload.png)
+
+A Data Submitter specifies the file ID, the URL to the Upload Contoller API, and the file path on the local file system using the CLI interface of the client (1.0). Internally, the CLI client translates the user-defined data into a request to the Upload Contoller API to obtain an S3 upload id (1.1). The client sends a request to the Upload Controller for each individual part of the uploaded file and recieves a presigned post for each individual part (1.2). The client reads multiple parts from the source file and uploads them as a stream using the pre-signed post (1.3). Once the upload has been completed the client sends a confirmation to the Upload Controller API (1.4).
+
+## CLI:
+
+The CLI commands remain the same.
+
+## UCS:
+
+The following Updates are performed in the Upload Controller Service:
+
+### Database:
+
+A new table "multipart_uploads" is added, containing the following columns:
+```
+- s3_upload_id (Primary key)
+- file_id (not unique)
+- upload_status
+```
+
+file_state is an Enum and can have the following values:
+
+```
+- pending
+- canceled
+- failed
+- uploaded
+- accepted
+- rejected
+```
+
+The column "upload_status" from then table "files" is removed.
+A column "s3_upload_id" (unique) is added to the table "files".
```

Comment:
I think this is not needed, because the "current" upload attempt can be reconstructed from the uploads table.